File: src/cmcandy/C_Language_Answers/_0024.cpp
#include <iostream>

using namespace ::std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution
{
public:
    ListNode *swapPairs(ListNode *head)
    {
        //基本情况
        if (head == NULL || head->next == NULL)
            return head;
        ListNode *pre = NULL, *now = head, *alf = head->next, *ntu = alf->next;

        while (ntu != NULL && ntu->next != NULL)
        {
            if (pre == NULL)
                head = alf;
            else
                pre->next = alf;
            //1->3
            now->next = ntu;
            //2->1
            alf->next = now;

            if (ntu == NULL || ntu->next == NULL)
                break;
            //next turn
            pre = now;
            now = now->next;
            alf = now->next;
            ntu = alf->next;
        }
        //最后一轮儿
        if (pre == NULL)
            head = alf;
        else
            pre->next = alf;
        now->next = ntu;
        alf->next = now;
        return head;
    }
};

File: src/cmcandy/Python_language_Answers/_0024.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:

        tmpHead = ListNode(next=head)
        tmp = tmpHead
        while tmp.next is not None and tmp.next.next is not None:
            tmp.next = self.swapPair(tmp.next)
            tmp = tmp.next.next
        return tmpHead.next

    def swapPair(self, head: ListNode):
        sec = head.next
        head.next = sec.next
        sec.next = head
        return sec